server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +29,142 @@ import org.http4s.Request
 import org.http4s.Status
 import org.http4s.dsl.io._
+import org.http4s.laws.discipline.arbitrary.genFiniteDuration
 import org.http4s.server.middleware.Throttle._
 import org.http4s.syntax.all._
+import org.scalacheck.effect.PropF.forAllF
 
 import scala.concurrent.duration._
 
 class ThrottleSuite extends Http4sSuite {
   test("LocalTokenBucket should contain initial number of tokens equal to specified capacity") {
-    val someRefillTime = 1234.milliseconds
-    val capacity = 5
-    val createBucket =
-      TokenBucket.local[IO](capacity, someRefillTime)
-
-    createBucket.flatMap { testee =>
-      val takeFiveTokens: IO[List[TokenAvailability]] =
-        (1 to 5).toList.traverse(_ => testee.takeToken)
-      val checkTokensUpToCapacity =
-        takeFiveTokens.map(tokens => tokens.contains(TokenAvailable))
-      (checkTokensUpToCapacity, testee.takeToken.map(_.isInstanceOf[TokenUnavailable]))
-        .mapN(_ && _)
-    }.assert
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val prog = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] =
+          (1 to 5).toList.traverse(_ => testee.takeToken)
+        val checkTokensUpToCapacity =
+          takeFiveTokens.map(tokens => tokens.contains(TokenAvailable))
+        (checkTokensUpToCapacity, testee.takeToken.map(_.isInstanceOf[TokenUnavailable]))
+          .mapN(_ && _)
+      }.assert
+
+      TestControl.executeEmbed(prog)
+    }
   }
 
   test("LocalTokenBucket should add another token at specified interval when not at capacity") {
-    val ctx = TestContext()
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val exceeded = someRefillTime + 1.millisecond
 
-    val capacity = 1
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+      val capacity = 1
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
 
-    val takeTokenAfterRefill = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(101.milliseconds) *>
-        testee.takeToken
-    }
+      val takeTokenAfterRefill = createBucket.flatMap { testee =>
+        testee.takeToken *> IO.sleep(exceeded) *>
+          testee.takeToken
+      }
 
-    takeTokenAfterRefill
-      .map { result =>
-        ctx.advanceAndTick(101.milliseconds)
-        result
+      TestControl.execute(takeTokenAfterRefill).flatMap { control =>
+        for {
+          _ <- control.results.assertEquals(None)
+          _ <- control.tick
+          _ <- control.results.assertEquals(None)
+          interval <- control.nextInterval
+          _ <- control.advanceAndTick(interval)
+          _ <- control.results.assertEquals(
+            Option(Outcome.succeeded[Id, Throwable, TokenAvailability](TokenAvailable.pure[Id]))
+          )
+        } yield ()
```

Comment:
Can we rewrite this to record a timestamp when `takeTokenAfterRefill` completes and assert that it's what we expect? And then we'd only need `executeEmbed` instead of the full control machinery.

Reply:
Wouldn't that be more imprecise than this implementation? If we take a timestamp we're saying that the new token was taken sometime up to that moment in time - with what we got now we're establishing that up to the interval it couldn't happen, and only after that the fiber was able to pick up the token. Thoughts?

Reply:
Hmm, not sure. For the record I could be misunderstanding what this test control is actually testing :)

> If we take a timestamp we're saying that the new token was taken sometime up to that moment in time

The timestamp would record the exact moment that the fiber continued, not before nor after.

> with what we got now we're establishing that up to the interval it couldn't happen, and only after that the fiber was able to pick up the token

The fact that the timestamp is not less than that interval would prove that up until then it couldn't happen.

Reply:
Same caveats apply to me: I've just learned about `TestControl`!

> The timestamp would record the exact moment that the fiber continued, not before nor after.
So you mean emitting a timestamp out of the test code execution flow? Hmmm, I think I'm seeing what you're getting at. OK, let's give it a spin